share/doc/wake/basic_wake_tutorial.md

```diff
@@ -0,0 +1,809 @@
+```
+def lex s = s | tokenize ` `
+```
+
+* Doubles every element in the list
+```
+def doublist n = match n 
+  Nil = Nil
+  h,t = 2*h , doublist t
+```
+
+* Increment the given list elements by 1
+```
+def inclist n = match n
+  Nil = Nil
+  h,t = (h+1), inclist t
+```
+
+* Doubles every element in the list using map function
+```
+def doublistmap n = n | map (\n 2*n)
+```
+
+* Increment every element in the list using map function
+```
+def inclistmap n = n | map (\n n+1) 
+```
+
+* Returns a list with the leading spaces removed
+```
+def dropspace n = match n
+  Nil = Nil
+  h,t = if (h ==* " ") then dropspace t else (h,t)
+```
+* Returns just the leading spaces
+```
+def takespace n = match n 
+  Nil = Nil
+  h,t = if (h ==* " ") then h,takespace t else Nil
+```
+
+* Concantenates every list of string with "io"
+```
+def map1 n = n | map (\s "{s}io")
+```
+
+* Creates a list (list Integers) from a given list of integers
+```
+def map2 n = n | map (\x x,Nil)
+```
+
+* Outputs the first element from the list (list Integers)
+```
+def map3 n = n | map (\x x | head | getOrElse 0)
+```
+
+* Creates a list of strings by talking the last character of a string
+```
+def map4 n = n | map (\x last x)
+```
+
+* Multiplies every element of the list by 3
+```
+def ftrl n = n | map (\x 3*x)
+```
+
+* Takes the first character from the list of strings and creates a new list
+```
+def fhel n = n | map (\x first x)
+```
+
+* removes first 2 characters from the list of strings and creates a new list
+```
+def tailstring n = n | explode | drop 1 |cat
+def fttl n = n | map (\x tailstring x)
+```
+
+* Prepend and Appends every element in a list of string by "s" "m"
+```
+def fsml n = n | map (\x "s{x}m")
+```
+
+### Creating your own Data types
+* Example shows how to create a data type of our own
+```
+global data Direction =
+  North
+  South
+  East
+  West
+```
+
+* Example shows the data type Money which can be in the form of Cash or the Cheque.
+```
+global data Money =
+  Cash Integer
+  Cheque (Pair String Double)
+```
+```
+def amount  = Cash 50
+def chequeamount = Cheque (Pair "HDFC" 350.00)
+```
+
+* Try out these and record the output
+```
+def queuePut x r = foldl (\a\b b,a)x r
+def queuePut1 r = foldl (\a\b b,a) Nil r
+```
+
+* Try out these examples
+```
+def listexample1 l e = l ++ (e,Nil)
+def listexample2 l = l | head | getOrElse 0 
+def listexample3 l = l | tail
+```
+`wake -x '("abc","def",Nil) | map(explode)|map ( head)'`
+
+* Find the sum of elements in a list using foldr function. Here n refers to the deafult value which the function returns if the list l is empty(Nil).
+```
+def suma l n= l | foldr(_+_)n
+```
+
+* Finds the mean in the given list of integers
+```
+def mean l = l | foldr(_+_)0 | (\x x/ len l)
+```
+
+* Finds the median in the given list of integers
+```
+def median l = 
+  def length = if (len l %2 ==0) then ((len l) /2)-1 else ((len l)/2)
+  l | drop length | head
+```
+
+#### Mutually Recursive Functions: Example
```

Comment:
what is the point of this section? why is it important for people to know?

Reply:
Just an example of how mutually recursive functions being called